File: 100/166 - Fraction to Recurring Decimal.java
import java.util.HashMap;

public class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        StringBuilder ret;
        long n, d;
        HashMap<Long, Long> hm;
        
        if (numerator == 0)
            return "0";
        
        ret = new StringBuilder();
        if ((numerator > 0) ^ (denominator > 0))
            ret.append('-');
        
        n = Math.abs((long)numerator);
        d = Math.abs((long)denominator);
        ret.append(n / d);
        n %= d;
        
        if (n == 0) 
            return ret.toString();
        else {
            ret.append('.');
            hm = new HashMap<Long, Long>();
        }
        
        while (n != 0) {
            if (hm.get(n) != null) {
                ret.insert(Math.toIntExact(hm.get(n)), '(');
                ret.append(')');
                break;
            }
            
            hm.put(n, (long)ret.length());
            n *= 10;
            ret.append(n / d);
            n %= d;
        }
        
        return ret.toString();
    }
}

File: csharp/166.fraction-to-recurring-decimal.cs
/*
 * @lc app=leetcode id=166 lang=csharp
 *
 * [166] Fraction to Recurring Decimal
 */

using System.Text;
using System.Collections.Generic;

public class Solution {
    public string FractionToDecimal(int numerator, int denominator) {
        StringBuilder sb = new StringBuilder();
        if (numerator != 0 && (numerator < 0 ^ denominator < 0)) {
            sb.Append('-');
        }
        long n = Math.Abs((long) numerator), d = Math.Abs((long) denominator);
        sb.Append(n / d);
        n = n % d;
        if (n == 0) {
            return sb.ToString();
        }
        IDictionary<long, int> map = new Dictionary<long, int>();
        sb.Append('.');
        while (!map.ContainsKey(n)) {
            map[n] = sb.Length;
            n *= 10;
            sb.Append(n / d);
            n = n % d;
            if (n == 0) {
                return sb.ToString();
            }
        }
        sb.Insert(map[n], '(');
        sb.Append(')');
        return sb.ToString();
    }
}

